### ISSUE TITLE
Roberta pre-training using torch-xla fails on GPU

### ISSUE THREAD

**Issue by hannanjgaws** [facebookresearch/fairseq#4198]
## 🐛 Bug

I am trying to pre-train Roberta base using torch-xla on GPU. I am following the [Pretraining RoBERTa using your own data tutorial](https://github.com/pytorch/fairseq/blob/main/examples/roberta/README.pretraining.md) to preprocess the WikiText-103 dataset and train RoBERTa base using 8 NVIDIA A100 GPUs. I added `common.tpu=True` to enable the XLA flow. 

The first training epoch starts, but then hangs indefinitely. I expect that using the XLA flow should work on GPU. 

### To Reproduce

Steps to reproduce the behavior (**always include the command you ran**):

1. Preprocess the WikiText-103 dataset following https://github.com/pytorch/fairseq/blob/main/examples/roberta/README.pretraining.md#1-preprocess-the-data
```
mkdir $HOME/pytorch-tutorial-data
cd $HOME/pytorch-tutorial-data

wget https://s3.amazonaws.com/research.metamind.io/wikitext/wikitext-103-raw-v1.zip
unzip wikitext-103-raw-v1.zip

mkdir -p gpt2_bpe
wget -O gpt2_bpe/encoder.json https://dl.fbaipublicfiles.com/fairseq/gpt2_bpe/encoder.json
wget -O gpt2_bpe/vocab.bpe https://dl.fbaipublicfiles.com/fairseq/gpt2_bpe/vocab.bpe

for SPLIT in train valid test; do \
    python -m examples.roberta.multiprocessing_bpe_encoder \
        --encoder-json gpt2_bpe/encoder.json \
        --vocab-bpe gpt2_bpe/vocab.bpe \
        --inputs wikitext-103-raw/wiki.${SPLIT}.raw \
        --outputs wikitext-103-raw/wiki.${SPLIT}.bpe \
        --keep-empty \
        --workers 60; \
done

wget -O gpt2_bpe/dict.txt https://dl.fbaipublicfiles.com/fairseq/gpt2_bpe/dict.txt

fairseq-preprocess \
    --only-source \
    --srcdict gpt2_bpe/dict.txt \
    --trainpref wikitext-103-raw/wiki.train.bpe \
    --validpref wikitext-103-raw/wiki.valid.bpe \
    --testpref wikitext-103-raw/wiki.test.bpe \
    --destdir data-bin/wikitext-103 \
    --workers 60
```

2. Set the following envvars to specify the data directory and use all 8 GPUs
```
export DATA_DIR=${HOME}/pytorch-tutorial-data/data-bin/wikitext-103
export GPU_NUM_DEVICES=8
```
3. Apply the following patch to FairSeq to fix an OOM issue:
```
diff --git a/fairseq/dataclass/configs.py b/fairseq/dataclass/configs.py
index 75407389..cbddcb75 100644
--- a/fairseq/dataclass/configs.py
+++ b/fairseq/dataclass/configs.py
@@ -4,6 +4,7 @@
 # LICENSE file in the root directory of this source tree.
 
 import sys
+import os
 from dataclasses import _MISSING_TYPE, dataclass, field
 from typing import Any, List, Optional
 
@@ -243,13 +244,13 @@ class CommonConfig(FairseqDataclass):
 @dataclass
 class DistributedTrainingConfig(FairseqDataclass):
     distributed_world_size: int = field(
-        default=max(1, torch.cuda.device_count()),
+        default=os.environ.get('GPU_NUM_DEVICES', 1),
         metadata={
             "help": "total number of GPUs across all nodes (default: all visible GPUs)"
         },
     )
     distributed_num_procs: Optional[int] = field(
-        default=max(1, torch.cuda.device_count()),
+        default=os.environ.get('GPU_NUM_DEVICES', 1),
         metadata={
             "help": "total number of processes to fork (default: all visible GPUs)"
         },
@@ -354,7 +355,7 @@ class DistributedTrainingConfig(FairseqDataclass):
         default=3, metadata={"help": "Local SGD allreduce frequency"}
     )
     nprocs_per_node: int = field(
-        default=max(1, torch.cuda.device_count()),
+        default=os.environ.get('GPU_NUM_DEVICES', 1),
         metadata={
             "help": "number of GPUs in each node. An allreduce operation across GPUs in "
             "a node is very fast. Hence, we do allreduce across GPUs in a node, "
```

4.  Run the Roberta base training command. Add `common.tpu=True` to enable to XLA flow.
```
cd fairseq

fairseq-hydra-train -m --config-dir examples/roberta/config/pretraining \
--config-name base task.data=$DATA_DIR common.tpu=True
```

The first training epoch starts, but hangs indefinitely.

#### Code sample

Code described above.

### Expected behavior

I expect that I should be able to use the TPU torch-xla flow to pre-train Roberta base on GPU. 

### Environment

 - fairseq Version (e.g., 1.0 or main): fairseq==1.0.0a0+4a7835b
 - PyTorch Version (e.g., 1.0): torch==1.10.1, torch-xla==1.10, torchvision==0.11.2
 - OS (e.g., Linux): Amazon Linux 2
 - How you installed fairseq (`pip`, source): from source
 - Build command you used (if compiling from source):
  ```
git clone https://github.com/pytorch/fairseq
cd fairseq
pip install --editable ./
cd ~/
 ```
 - Python version: Python 3.7.12
 - CUDA/cuDNN version: V11.2.152
 - GPU models and configuration: 8x A100
 - Any other relevant information:

**Comment by hannanjgaws** [comment]
It turns out the continuous recompilation was happening because FairSeq was feeding the XLA model dynamic inputs. The continuous recompilation can be fixed by padding the inputs to a max sequence length. Pre-training Roberta base using torch-xla + fairseq is now successful on 8 GPUs. 

For posterity's sake, here's how I debugged the issue:
* I used `PT_XLA_DEBUG=1` to determine that re-compilation was happening nearly every training step. For example, at the 493 step, I saw the following profiler output: `pt-xla-profiler: CompileTime too frequent: 429 counts during 493 steps`
* I used `XLA_SAVE_TENSORS_FILE` to save the graph IRs that we being compiled. A diff on different graphs revealed the input shape was changing.
* Lastly, I confirmed that the shape of the `src_tokens` input into `RobertaModel.forward` was changing each iteration. Applying the following patch fixed this issue and enabled successful training:
```
diff --git a/fairseq/data/data_utils.py b/fairseq/data/data_utils.py
index 9cd88ca7..82d05f99 100644
--- a/fairseq/data/data_utils.py
+++ b/fairseq/data/data_utils.py
@@ -40,7 +40,7 @@ def collate_tokens(
     eos_idx=None,
     left_pad=False,
     move_eos_to_beginning=False,
-    pad_to_length=None,
+    pad_to_length=512,
     pad_to_multiple=1,
     pad_to_bsz=None,
 ):
```

I'll close this ticket since the original issue (hanging training) is fixed.

**Comment by salrowili** [comment]
@hannanjgaws  You are my hero. I have been struggling to pre-train BART with fairseq because of XLA dynamic shape issue. I was really surprised that it turns out to be linked to this variable which was not provided to the training arguments. In addition, the bucket cap argument is another option to consider to accelerate training on TPU. try smaller number like 5 and see how it performs compared to 25.

**Comment by salrowili** [comment]
Do you think this is the same idea of --input_shapes 256x64 512x32 in this example https://cloud.google.com/tpu/docs/tutorials/transformer-pytorch

if so can those numbers for input_shape help us here?
